### DISCUSSION TITLE
stm32f407 high speed USB

### DISCUSSION THREAD

**Discussion by fizzyade** [hathach/tinyusb#1735] [Q&A]
Hi,

I'm trying to test tinyUSB on the stm32f407 with a high speed phy (USB3300), I've been reading up and it seems that tinyUSB is more performant than the HAL provided USB implementation.

I have been using the STM32 HAL USB library with a vendor specific bulk interface and it is working, but I'm wanting to use tinyUSB to see if I can get even faster transfers, I have verified that the HAL implementation is working 100% with a hardware USB protocol analyser.

I've tried to get tinyUSB working, but it's locking up during initialisation, I've tried this with other libraries and there seems to be. very specific requirements for enabling the PHY clock, if it's initialised in the wrong order then the STM32 core completely locks up, I've read about this behaviour with external PHY's elsewhere.

Have you tried tinyUSB on a f407 with an external PHY?  The f407 examples supplied only use the internal FS PHY and I've tried to modify based on another stm32 port that uses a HS phy, but I'm hitting the issue with the core locking up.

**Comment by hathach** [comment]
no I haven't tried f407 specially with external PHY since I don't have any of that boards. But tinyusb can run well with h743 + external PHY. I guess it is mostly configuration. You could take a peek at h743eval bsp code 

https://github.com/hathach/tinyusb/blob/master/hw/bsp/stm32h7/boards/stm32h743eval/board.h

Also make sure you set `CFG_TUD_MAX_SPEED` to highspeed an init correct rhport = 1 (roothub port). ( 0 for FS, 1 for HS for stm mcu). Also attached debug log as txt (LOG=2) here, it can contain some hint on where the code is locked up.
